compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -2389,15 +2505,36 @@ fn convert_array_expression_to_slice(
 
 impl BinaryTypeOperator {
     /// Perform the actual rust numeric operation associated with this operator
-    // TODO(https://github.com/noir-lang/noir/pull/6137): the Kind is included
-    // since it'll be needed for size checks
-    pub fn function(self, a: u32, b: u32, _kind: &Kind) -> Option<u32> {
-        match self {
-            BinaryTypeOperator::Addition => a.checked_add(b),
-            BinaryTypeOperator::Subtraction => a.checked_sub(b),
-            BinaryTypeOperator::Multiplication => a.checked_mul(b),
-            BinaryTypeOperator::Division => a.checked_div(b),
-            BinaryTypeOperator::Modulo => a.checked_rem(b),
+    pub fn function(self, a: FieldElement, b: FieldElement, kind: &Kind) -> Option<FieldElement> {
+        match kind.follow_bindings().integral_maximum_size() {
+            None => match self {
+                BinaryTypeOperator::Addition => Some(a + b),
+                BinaryTypeOperator::Subtraction => Some(a - b),
+                BinaryTypeOperator::Multiplication => Some(a * b),
+                BinaryTypeOperator::Division => {
+                    if b == FieldElement::zero() {
+                        None
+                    } else {
+                        Some(a / b)
+                    }
+                }
+                BinaryTypeOperator::Modulo => None,
+            },
+            Some(maximum_size) => {
+                let maximum_size = maximum_size.to_i128();
+                let a = a.to_i128();
+                let b = b.to_i128();
+
+                let result = match self {
+                    BinaryTypeOperator::Addition => (a + b) % maximum_size,
+                    BinaryTypeOperator::Subtraction => (a - b) % maximum_size,
+                    BinaryTypeOperator::Multiplication => (a * b) % maximum_size,
+                    BinaryTypeOperator::Division => (a.checked_div(b)?) % maximum_size,
+                    BinaryTypeOperator::Modulo => a.checked_rem(b)?,
```

Comment:
Looks like we wrap now instead of error on overflow - I think we should keep the error